### PR TITLE
refactor DAG and DAG consumers to support >2 Listeners

### PR DESCRIPTION
**Pull request by skriss**
Refactors DAG internals to support
any number of Listeners to prepare
for future changes to support many
Gateway Listeners.

Updates #4960.

I think this is in good shape, but as the design doc hasn't been merged yet leaving as draft for now.

### CODE REVIEW

**Comment by skriss** [comment]
Since there's no functional change here, not including a release note. I'll include a comprehensive one with the next PR to support >2 Listeners for Gateways.

File: internal/dag/dag.go [hunk 1]
```diff
@@ -59,10 +59,13 @@ type DAG struct {
...
-	VirtualHosts       map[string]*VirtualHost
-	SecureVirtualHosts map[string]*SecureVirtualHost
-	ExtensionClusters  []*ExtensionCluster
+	Listeners         map[string]*Listener
+	ExtensionClusters []*ExtensionCluster
+
+	// Set this to true if Contour is configured with a Gateway
+	// and Listeners are derived from the Gateway's Listeners, or
+	// false otherwise.
+	HasNonStaticListeners bool
```
Comment (sunjayBhatia) [inline internal/dag/dag.go:68]: tiny naming nit: could use a name w/o negation to make situations like `!HasNonStaticListeners` read a little better (e.g. `HasDynamicListeners`)

**Review by sunjayBhatia — APPROVED** [review]
this looks great! took a look at this and how its used in your other PR 👍🏽

File: internal/dag/accessors.go [hunk 2]
```diff
@@ -121,14 +121,14 @@ func externalName(svc *v1.Service) string {
 // GetSecureVirtualHost returns the secure virtual host in the DAG that
 // matches the provided name, or nil if no matching secure virtual host
 // is found.
-func (d *DAG) GetSecureVirtualHost(hostname string) *SecureVirtualHost {
-	return d.SecureVirtualHosts[hostname]
+func (d *DAG) GetSecureVirtualHost(listener, hostname string) *SecureVirtualHost {
+	return d.Listeners[listener].svhostsByName[hostname]
```
[the remark this answers is not in the document]
Comment (skriss) [reply internal/dag/accessors.go:125]: Good call, yeah this is one of those things where logically the listener _should_ always be there if we're asking to get/ensure a virtual host for it, but hard to guarantee that, so should probably be defensive.

[the remark this answers is not in the document]
Comment (skriss) [reply internal/dag/accessors.go:125]: I'll plan to address this in the next PR when we add the dynamic listener handling

**Review by sunjayBhatia — APPROVED** [review]

**Pull request merged**